packages/core/src/collaboration.ts

```diff
@@ -13,175 +13,41 @@
 // limitations under the License.
 //
 
-import { Doc, Ref } from './classes'
-
-/**
- * Identifier of the collaborative document holding collaborative content.
- *
- * Format:
- * {documentId}:{versionId}:{lastVersionId}
- * {documentId}:{versionId}
- *
- * Where:
- * - documentId is an identifier of the document in storage
- * - versionId is an identifier of the document version, HEAD for latest editable version
- * - lastVersionId is an identifier of the latest available version
- *
- * The collaborative document may contain one or more such sections chained with # (hash):
- * collaborativeDocId#collaborativeDocId#collaborativeDocId#...
- *
- * When collaborative document does not exist, it will be initialized from the first existing
- * document in the list.
- *
- * @public
- * */
-export type CollaborativeDoc = string & { __collaborativeDoc: true }
-
-/** @public */
-export type CollaborativeDocVersion = string | typeof CollaborativeDocVersionHead
+import type { Blob, Class, Doc, Ref } from './classes'
 
 /** @public */
-export const CollaborativeDocVersionHead = 'HEAD'
+export interface CollaborativeDoc {
+  objectClass: Ref<Class<Doc>>
+  objectId: Ref<Doc>
+  objectAttr: string
+}
 
 /** @public */
-export function makeCollaborativeDoc (
-  objectId: Ref<Doc>,
-  objectAttr?: string | undefined,
-  versionId?: string | undefined
+export function makeCollabId<T extends Doc, U extends keyof T> (
+  objectClass: Ref<Class<T>>,
+  objectId: Ref<T>,
+  objectAttr: Extract<U, string> | string
```

Comment:
Just curious, isn't `| string` removing the check for the attribute to be a valid key completely?

Reply:
Yep, but in editor it still works better than just string. I failed to do this properly because of many places where we process document in generic way when we don't have object type - in this case this utility function becomes unusable.